File: searchnsort/search_2D.py
'''
leetcode problem 74 - search a 2D matrix
'''

def binary_search(nums, target):
    l, e = 0, len(nums)-1

    while l <= e:
        m = (l+e)//2
        if nums[m] == target : return True
        elif nums[m] > target : e = m-1
        else : l = m+1

    return False

def searchMatrix(self, matrix, target):
    fr, lr, lc = 0, len(matrix)-1, len(matrix[0])-1

    while fr <= lr:
        mr = (fr+lr)//2
        if(matrix[mr][0] <= target and matrix[mr][lc] >= target):
            return binary_search(matrix[mr], target)
        elif(matrix[mr][lc] < target): fr = mr+1
        else: lr = mr-1

    return False

File: searchnsort/search_2D.cpp
//leetcode problem 74 - search a 2D matrix

bool searchMatrix(vector<vector<int>>& a, int x) {
        int n=a.size();
        if(n==0) return 0;
        int m=a[0].size();
        int l=0,h=n*m-1;
        while(l<=h) {
            int mid=(l+h)/2;
            int r=mid/m;
            int c=mid%m;
            if(a[r][c]==x) return 1;
            if(a[r][c]>x) h=mid-1;
            else l=mid+1;
        }
        return 0;
    }
